File: backend/app/schemas/gamification.py
from __future__ import annotations

import enum
from datetime import datetime
from typing import Any, Dict, List

from pydantic import BaseModel, ConfigDict, Field


class LeaderboardMetric(str, enum.Enum):
    score = "score"
    completion_rate = "completion_rate"
    engagement = "engagement"
    quiz_accuracy = "quiz_accuracy"


class BadgeBase(BaseModel):
    name: str
    description: str | None = None
    icon: str | None = None
    criteria: Dict[str, Any] | None = None


class BadgeCreate(BadgeBase):
    pass


class BadgeUpdate(BaseModel):
    name: str | None = None
    description: str | None = None
    icon: str | None = None
    criteria: Dict[str, Any] | None = None


class BadgeRead(BadgeBase):
    id: int
    organization_id: int
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class AchievementBase(BaseModel):
    name: str
    description: str | None = None
    icon: str | None = None
    criteria: Dict[str, Any] | None = None
    points_reward: float = 0.0


class AchievementCreate(AchievementBase):
    pass


class AchievementUpdate(BaseModel):
    name: str | None = None
    description: str | None = None
    icon: str | None = None
    criteria: Dict[str, Any] | None = None
    points_reward: float | None = None


class AchievementRead(AchievementBase):
    id: int
    organization_id: int
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class UserAchievementRead(BaseModel):
    id: int
    achievement: AchievementRead
    awarded_at: datetime

    model_config = ConfigDict(from_attributes=True)


class LevelDefinitionBase(BaseModel):
    name: str
    description: str | None = None
    threshold_points: float = Field(..., ge=0.0)


class LevelDefinitionCreate(LevelDefinitionBase):
    pass


class LevelDefinitionUpdate(BaseModel):
    name: str | None = None
    description: str | None = None
    threshold_points: float | None = Field(default=None, ge=0.0)


class LevelDefinitionRead(LevelDefinitionBase):
    id: int
    organization_id: int
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class UserProgressRead(BaseModel):
    user_id: int
    total_points: float
    current_level: LevelDefinitionRead | None = None
    next_level_threshold: float | None = None
    badges: List[BadgeRead] = Field(default_factory=list)
    achievements: List[UserAchievementRead] = Field(default_factory=list)

    model_config = ConfigDict(from_attributes=True)


class UserPointsUpdate(BaseModel):
    points_delta: float


class LeaderboardCreate(BaseModel):
    name: str
    description: str | None = None
    metric: LeaderboardMetric = LeaderboardMetric.score
    is_public: bool = True


class LeaderboardUpdate(BaseModel):
    name: str | None = None
    description: str | None = None
    metric: LeaderboardMetric | None = None
    is_public: bool | None = None


class LeaderboardEntryRead(BaseModel):
    id: int
    leaderboard_id: int
    user_id: int
    score: float
    rank: float
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class LeaderboardEntryWithUserRead(BaseModel):
    id: int
    leaderboard_id: int
    user_id: int
    user_name: str
    score: float
    rank: float
    user_level: str
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class LeaderboardRead(BaseModel):
    id: int
    name: str
    description: str | None = None
    metric: LeaderboardMetric
    is_public: bool
    entries: List[LeaderboardEntryRead] = Field(default_factory=list)

    model_config = ConfigDict(from_attributes=True)


class UserBadgeRead(BaseModel):
    id: int
    badge: BadgeRead
    awarded_at: datetime

    model_config = ConfigDict(from_attributes=True)


class GamificationSummaryRead(BaseModel):
    total_points: float
    current_level: LevelDefinitionRead | None = None
    next_level_threshold: float | None = None
    badges: List[BadgeRead] = Field(default_factory=list)
    achievements: List[UserAchievementRead] = Field(default_factory=list)

    model_config = ConfigDict(from_attributes=True)

File: frontend/types/gamification.ts
export type LeaderboardMetric = 'score' | 'completion_rate' | 'engagement' | 'quiz_accuracy';

export interface Badge {
  id: number;
  organization_id: number;
  name: string;
  description?: string;
  icon?: string;
  criteria?: Record<string, any>;
  created_at: string;
  updated_at: string;
}

export interface Achievement {
  id: number;
  organization_id: number;
  name: string;
  description?: string;
  icon?: string;
  criteria?: Record<string, any>;
  points_reward: number;
  created_at: string;
  updated_at: string;
}

export interface UserAchievement {
  id: number;
  achievement: Achievement;
  awarded_at: string;
}

export interface LevelDefinition {
  id: number;
  organization_id: number;
  name: string;
  description?: string;
  threshold_points: number;
  created_at: string;
  updated_at: string;
}

export interface GamificationSummary {
  total_points: number;
  current_level?: LevelDefinition | null;
  next_level_threshold?: number | null;
  badges: Badge[];
  achievements: UserAchievement[];
}

export interface BadgeCreate {
  name: string;
  description?: string;
  icon?: string;
  criteria?: Record<string, any>;
}

export interface BadgeUpdate {
  name?: string;
  description?: string;
  icon?: string;
  criteria?: Record<string, any>;
}

export interface AchievementCreate {
  name: string;
  description?: string;
  icon?: string;
  criteria?: Record<string, any>;
  points_reward: number;
}

export interface AchievementUpdate {
  name?: string;
  description?: string;
  icon?: string;
  criteria?: Record<string, any>;
  points_reward?: number;
}

export interface LevelDefinitionCreate {
  name: string;
  description?: string;
  threshold_points: number;
}

export interface LevelDefinitionUpdate {
  name?: string;
  description?: string;
  threshold_points?: number;
}

export interface LeaderboardCreate {
  name: string;
  description?: string;
  metric: LeaderboardMetric;
  is_public: boolean;
}

export interface LeaderboardUpdate {
  name?: string;
  description?: string;
  metric?: LeaderboardMetric;
  is_public?: boolean;
}

export interface LeaderboardEntry {
  id: number;
  leaderboard_id: number;
  user_id: number;
  score: number;
  rank: number;
  created_at: string;
  updated_at: string;
}

export interface LeaderboardEntryWithUser {
  id: number;
  leaderboard_id: number;
  user_id: number;
  user_name: string;
  score: number;
  rank: number;
  user_level: string;
  created_at: string;
  updated_at: string;
}

export interface Leaderboard {
  id: number;
  name: string;
  description?: string;
  metric: LeaderboardMetric;
  is_public: boolean;
  entries: LeaderboardEntry[];
}

export interface LeaderboardWithUsers {
  id: number;
  name: string;
  description?: string;
  metric: LeaderboardMetric;
  is_public: boolean;
  entries: LeaderboardEntryWithUser[];
}
